ir/src/constraints/list_comprehension.rs

```diff
@@ -0,0 +1,350 @@
+}
+
+/// Parses a list comprehension expression and creates an expression based on the index of the
+/// expression in the list comprehension.
+///
+/// Returns an error if there is an error while parsing the sub-expression.
+fn parse_lc_expr(
+    expression: &Expression,
+    iterable_context: &IterableContext,
+    symbol_table: &SymbolTable,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    match expression {
+        Expression::Elem(ident) => parse_elem(ident, expression, iterable_context, symbol_table, i),
+        Expression::NamedTraceAccess(named_trace_access) => parse_named_trace_access(
+            named_trace_access,
+            expression,
+            iterable_context,
+            symbol_table,
+            i,
+        ),
+        Expression::Add(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, iterable_context, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, iterable_context, symbol_table, i)?;
+            Ok(Expression::Add(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::Sub(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, iterable_context, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, iterable_context, symbol_table, i)?;
+            Ok(Expression::Sub(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::Mul(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, iterable_context, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, iterable_context, symbol_table, i)?;
+            Ok(Expression::Mul(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::Exp(lhs, rhs) => {
+            let lhs = parse_lc_expr(lhs, iterable_context, symbol_table, i)?;
+            let rhs = parse_lc_expr(rhs, iterable_context, symbol_table, i)?;
+            Ok(Expression::Exp(Box::new(lhs), Box::new(rhs)))
+        }
+        Expression::ListFolding(lf_type) => {
+            parse_list_folding(lf_type, expression, symbol_table, i)
+        }
+        _ => Ok(expression.clone()),
+    }
+}
+
+/// Parses an identifier in a list comprehension expression.
+///
+/// # Errors
+/// - Returns an error if the iterable is an identifier and that identifier does not correspond to
+///   a vector.
+/// - Returns an error if the iterable is an identifier but is not of a type in set:
+///   { TraceColumns, IntegrityVariable, PublicInput, RandomValuesBinding }.
+/// - Returns an error if the iterable is a slice and that identifier does not correspond to
+///   a vector.
+/// - Returns an error if the iterable is an identifier but is not of a type in set:
+///   { TraceColumns, IntegrityVariable, PublicInput, RandomValuesBinding }.
+fn parse_elem(
+    ident: &Identifier,
+    expression: &Expression,
+    iterable_context: &IterableContext,
+    symbol_table: &SymbolTable,
+    i: usize,
+) -> Result<Expression, SemanticError> {
+    let iterable = iterable_context.get(ident);
+    if let Some(iterable_type) = iterable {
+        match iterable_type {
```

Comment:
@bobbinth I think we can have symbols in a list comprehension expression that are not in the iterables of the list comprehension. For example, the following should be valid:
```
trace_columns:
    main: [c[3]]
let x = 1
let a = [b + x for b in c]
```
Here, when we are unrolling the list comprehension, we just leave x alone and return it if we don't find a corresponding iterable. So the above list comprehension unrolls into:
```
[c[0] + x, c[1] + x, c[2] + x]
```

Reply:
Ah - makes sense! I'd probably add some comments and maybe restructure the function to make it more clear.

For example, for `parse_elem()` do we really need to pass in `expression`? If I understood it correctly, passing both `ident` and `expression` is redundant. And the only reason we pass it in is to clone it at the end. But we could have just as easily done something like `Expression::Elem(ident)` right?

And the logic of the whole function could be:
1. If the `ident` is not in the `iterable_context`, then just return `Expression::Elem(ident)` (and I guess any invalid references will be caught later).
2. Otherwise, process the `ident` as we do now.